README.rst
The Module Build Service (MBS) for Modularity
=============================================

The MBS coordinates module builds and is responsible for a number of
tasks:

- Providing an interface for module client-side tooling via which module build
  submission and build state queries are possible.
- Verifying the input data (modulemd, RPM SPEC files and others) is available
  and correct.
- Preparing the build environment in the supported build systems, such as koji.
- Scheduling and building of the module components and tracking the build
  state.
- Emitting bus messages about all state changes so that other infrastructure
  services can pick up the work.


Change Log
==========

For a detailed change log, see |docs/CHANGELOG.rst|_.

.. |docs/CHANGELOG.rst| replace:: ``docs/CHANGELOG.rst``
.. _docs/CHANGELOG.rst: docs/CHANGELOG.rst

Contributing
============

For detailed information on how to contribute, see |docs/CONTRIBUTING.rst|_.

.. |docs/CONTRIBUTING.rst| replace:: ``docs/CONTRIBUTING.rst``
.. _docs/CONTRIBUTING.rst: docs/CONTRIBUTING.rst

Rebuild Strategies
==================

For detailed information on rebuild strategies, see |docs/REBUILD_STRATEGIES.rst|_.

.. |docs/REBUILD_STRATEGIES.rst| replace:: ``docs/REBUILD_STRATEGIES.rst``
.. _docs/REBUILD_STRATEGIES.rst: docs/REBUILD_STRATEGIES.rst

Supported build systems
=======================

Koji
----

Koji is the software that builds RPM packages from source tarballs and
SPEC files. It uses its own Mock to create chroot environments to
perform builds.

MBS comes with its own ``koji.conf`` config file which allows configuring
for your custom Koji instance(s).

Mock
----

Mock is a simple program that will build source RPMs inside a chroot. It
doesn't do anything terribly fancy other than populate a chroot with the
contents specified by a configuration file, then build any input SRPM(s)
in that chroot.

MBS is able to perform local module builds by directing local Mock.

MBS supports threaded Mock builds which utilizes performance and
significantly speeds up local module builds.

_`Client tooling`
=================

``mbs-manager``
---------------

This command controls the MBS instance itself.

There are subcommands for running the MBS server, performing database
migrations, generating certificates and submitting local module
builds. For more info, there's an existing help available.

Client-side API
===============

The MBS implements a RESTful interface for module build submission and state
querying. Not all REST methods are supported. See below for details. For client
tooling which utilizes the API, please refer to `Client tooling`_ section.

Module build submission
-----------------------

Module submission is done via posting the modulemd SCM URL.

::

    POST /module-build-service/1/module-builds/

::

    {
        "scmurl": "https://src.fedoraproject.org/modules/foo?#f1d2d2f924e986ac86fdf7b36c94bcdf32beec15",
        "branch": "master"
    }

The response, in case of a successful submission, would include the task ID.

::

    HTTP 201 Created

::

    {
        "id": 42,
        "state": "wait",
        ...
    }

Options:

- ``buildrequire_overrides`` - the buildrequires to override the modulemd with. The overrides must
  be to existing buildrequires on the modulemd. The expected format is
  ``{'platform': ['f28', 'f29']}``.
- ``modulemd`` - a string for submitting a YAML modulemd file as a parameter in the JSON data as
  an alternative to sending it in a ``multipart/form-data`` request. Only allowed if
  ``scratch`` is ``True`` or if the MBS setting ``YAML_SUBMIT_ALLOWED`` is ``True``.
- ``module_name`` - a string to use as the module name if a scratch build is requested and the
  YAML modulemd is submitted using the ``modulemd`` parameter.
- ``module_stream`` - a string to use as the stream name if a scratch build is requested and the
  YAML modulemd is submitted using the ``modulemd`` parameter.
- ``rebuild_strategy`` - a string of the desired rebuild strategy (affects what components get
  rebuilt). For the available options, please look at the "Rebuild Strategies" section below.
- ``require_overrides`` - the requires to override the modulemd with. The overrides must be to
  existing requires on the modulemd. The expected format is ``{'platform': ['f28', 'f29']}``.
- ``reuse_components_from`` - the ID or NSVC of the module build to reuse components from. If it's
  not set, MBS will try to find a compatible module build to reuse components from.
- ``scratch`` - a boolean indicating if a scratch module build should be performed.
  Only allowed to be ``True`` if the MBS setting ``MODULES_ALLOW_SCRATCH`` is ``True``.
- ``srpms`` - an optional list of Koji upload URLs of SRPMs to include in a module scratch build.
  Only allowed if ``scratch`` is ``True``.
- ``yaml`` - a string of the input file when submitting a YAML modulemd file directly in a
  ``multipart/form-data`` request. Only allowed if ``scratch`` is ``True`` or if the MBS
  setting ``YAML_SUBMIT_ALLOWED`` is ``True``. The basename of the file will be used as
  the module name.


Module build state query
------------------------

Once created, the client can query the current build state by requesting the
build task's URL.

::

    GET /module-build-service/1/module-builds/1042

The response, if the task exists, would include various pieces of information
about the referenced build task.

::

    HTTP 200 OK

::

    {
      "id": 1042,
      "koji_tag": "module-f8c7dcdcc884bf1d",
      "name": "cloud-init",
      "owner": "karsten",
      "scmurl": "https://src.fedoraproject.org/modules/cloud-init?#d5fc9ab58f359b618e67ebdd0c7b143962242546",
      "state": 5,
      "state_name": "ready",
      "state_reason": null,
      "stream": "master",
      "tasks": {
        "rpms": {
          "cloud-init": {
            "nvr": "cloud-init-0.7.9-9.module_f8c7dcdc",
            "state": 1,
            "state_reason": "",
            "task_id": 22264880
          },
          "module-build-macros": {
            "nvr": "module-build-macros-0.1-1.module_f8c7dcdc",
            "state": 1,
            "state_reason": "",
            "task_id": 22264426
          },
          "pyserial": {
            "nvr": "pyserial-3.1.1-5.module_f8c7dcdc",
            "state": 1,
            "state_reason": "",
            "task_id": 22264727
          },
          ...
        }
      },
      "time_completed": "2017-10-05T11:58:44Z",
      "time_modified": "2017-10-05T11:58:58Z",
      "time_submitted": "2017-10-05T11:37:39Z",
      "version": "20171005113458"
    }

The response includes:

- ``id`` - the ID of the module build.
- ``koji_tag`` - the Koji tag the component builds are tagged in.
- ``name`` - the name of the module.
- ``owner`` - the username of the owner or person who submitted the module build.
- ``scmurl`` - the source control URL used to build the module.
- ``state`` - the numerical state of the module build.
- ``state_name`` - the named state of the module build. See the section called
  `Module Build States`_ for more information.
- ``state_reason`` - the reason why the module build is in this state. This is useful
  when the build fails.
- ``stream`` - the module's stream.
- ``tasks`` - a dictionary of information about the individual component builds.
- ``time_completed`` - Zulu ISO 8601 timestamp of when the module build completed.
- ``time_modified`` - Zulu ISO 8601 timestamp of when the module build was last modified.
- ``time_submitted`` - Zulu ISO 8601 timestamp of when the module build was submitted.
- ``version`` - the module build's version.


Listing all module builds
-------------------------

The list of all tracked builds and their states can be obtained by
querying the "module-builds" resource.
There are a number of configurable GET parameters to change how the
module builds are displayed. These parameters are:

- ``verbose`` - Shows the builds with additional detail such as the modulemd
  and state trace (i.e. ``verbose=True``). This value defaults to ``False``.
- ``short`` - Shows the builds with a minimum amount of information
  (i.e. ``short=True``). This value defaults to ``False``.
- ``page`` - Specifies which page should be displayed (e.g. ``page=3``). This
  value defaults to 1.
- ``per_page`` - Specifies how many items per page should be displayed
  (e.g. ``per_page=20``). This value defaults to 10.
- ``order_by`` - a database column to order the API by in ascending order. Multiple can be provided.
- ``order_desc_by`` - a database column to order the API by in descending order. Multiple can be
  provided. This defaults to ``id``.

An example of querying the "module-builds" resource with the "per_page" and the "page"
parameters::

    GET /module-build-service/1/module-builds/?per_page=2&page=1

::

    HTTP 200 OK

::

    {
      "items": [
        {
          "id": 124,
          "koji_tag": "module-de66baf89b40367c",
          "name": "testmodule",
          "owner": "mprahl",
          "scmurl": "https://src.fedoraproject.org/modules/testmodule.git?#86d9cfe53d20118d863ae051641fc3784d91d981",
          "state": 5,
          "state_name": "ready",
          "state_reason": null,
          "stream": "master",
          "tasks": {
            "rpms": {
              "ed": {
                "nvr": "ed-1.14.1-4.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22267993
              },
              "mksh": {
                "nvr": "mksh-56b-1.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22268059
              }
            }
          },
          "time_completed": "2017-10-05T18:45:56Z",
          "time_modified": "2017-10-05T18:46:10Z",
          "time_submitted": "2017-10-05T18:34:39Z",
          "version": "20171005183359"
        },
        {
          "id": 123,
          "koji_tag": "module-4620ad476f3d2b5c",
          "name": "testmodule",
          "owner": "mprahl",
          "scmurl": "https://src.fedoraproject.org/modules/testmodule.git?#373bb6eccccbfebbcb222a2723e643e7095c7973",
          "state": 5,
          "state_name": "ready",
          "state_reason": null,
          "stream": "master",
          "tasks": {
            "rpms": {
              "ed": {
                "nvr": "ed-1.14.1-4.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22267993
              },
              "mksh": {
                "nvr": "mksh-56b-1.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22268059
              }
            }
          },
          "time_completed": "2017-10-05T18:45:50Z",
          "time_modified": "2017-10-05T18:46:01Z",
          "time_submitted": "2017-10-05T18:24:09Z",
          "version": "20171005182359"
        }
      ],
      "meta": {
        "first": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?per_page=2&page=1",
        "last": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?per_page=2&page=340",
        "next": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?per_page=2&page=2",
        "page": 1,
        "pages": 60,
        "per_page": 2,
        "prev": null,
        "total": 120
      }
    }


An example of querying the "module-builds" resource with the "verbose", "per_page", and the "page"
parameters::

    GET /module-build-service/1/module-builds/?per_page=2&page=1?verbose=true

::

    HTTP 200 OK

::

    {
      "items": [
        {
          "base_module_buildrequires": [
            {
              "context": "00000000",
              "name": "platform",
              "stream": "f29",
              "stream_version": 290000,
              "version": "5"
            }
          ],
          "component_builds": [
            57047,
            57048
          ],
          "id": 124,
          "koji_tag": "module-de66baf89b40367c",
          "modulemd": "...."
          "name": "testmodule",
          "owner": "mprahl",
          "reused_module_id": 121,
          "scmurl": "https://src.fedoraproject.org/modules/testmodule.git?#86d9cfe53d20118d863ae051641fc3784d91d981",
          "state": 5,
          "state_name": "ready",
          "state_reason": null,
          "state_trace": [
            {
              "reason": null,
              "state": 1,
              "state_name": "wait",
              "time": "2017-10-05T18:34:50Z"
            },
            ...
          ],
          "state_url": "/module-build-service/1/module-builds/1053",
          "stream": "master",
          "tasks": {
            "rpms": {
              "ed": {
                "nvr": "ed-1.14.1-4.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22267993
              },
              "mksh": {
                "nvr": "mksh-56b-1.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22268059
              }
            }
          },
          "time_completed": "2017-10-05T18:45:56Z",
          "time_modified": "2017-10-05T18:46:10Z",
          "time_submitted": "2017-10-05T18:34:39Z",
          "version": "20171005183359"
        },
        {
          "component_builds": [
            57045,
            57046
          ],
          "id": 123,
          "koji_tag": "module-4620ad476f3d2b5c",
          "modulemd": "...."
          "name": "testmodule",
          "owner": "mprahl",
          "scmurl": "https://src.fedoraproject.org/modules/testmodule.git?#373bb6eccccbfebbcb222a2723e643e7095c7973",
          "state": 5,
          "state_name": "ready",
          "state_reason": null,
          "state_trace": [
            {
              "reason": null,
              "state": 1,
              "state_name": "wait",
              "time": "2017-10-05T18:24:19Z"
            },
            ...
          ],
          "state_url": "/module-build-service/1/module-builds/1052",
          "stream": "master",
          "tasks": {
            "rpms": {
              "ed": {
                "nvr": "ed-1.14.1-4.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22267993
              },
              "mksh": {
                "nvr": "mksh-56b-1.module_d2a2f5c8",
                "state": 1,
                "state_reason": "Reused component from previous module build",
                "task_id": 22268059
              }
            }
          },
          "time_completed": "2017-10-05T18:45:50Z",
          "time_modified": "2017-10-05T18:46:01Z",
          "time_submitted": "2017-10-05T18:24:09Z",
          "version": "20171005182359"
        }
      ],
      "meta": {
        "first": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?verbose=true&per_page=2&page=1",
        "last": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?verbose=true&per_page=2&page=340",
        "next": "http://mbs.fedoraproject.org/module-build-service/1/module-builds/?verbose=true&per_page=2&page=2",
        "page": 1,
        "pages": 120,
        "per_page": 2,
        "prev": null,
        "total": 60
      }
    }

Filtering module builds
-----------------------

The module builds can be filtered by a variety of GET parameters. Some of these
parameters include:

- ``base_module_br`` - the name:stream:version:context of a base module the module buildrequires
- ``base_module_br_context`` - the context of a base module the module buildrequires
- ``base_module_br_name`` - the name of a base module the module buildrequires
- ``base_module_br_stream`` - the stream of a base module the module buildrequires
- ``base_module_br_stream_version`` - the stream version of a base module the module buildrequires
- ``base_module_br_stream_version_lte`` - less than or equal to the stream version of a base module
  the module buildrequires
- ``base_module_br_stream_version_gte`` - greater than or equal to the stream version of a base
  module the module buildrequires
- ``base_module_br_version`` - the version of a base module the module buildrequires
- ``batch``
- ``cg_build_koji_tag``
- ``completed_after`` - Zulu ISO 8601 format e.g. ``completed_after=2016-08-23T09:40:07Z``
- ``completed_before`` - Zulu ISO 8601 format e.g. ``completed_before=2016-08-22T09:40:07Z``
- ``koji_tag``
- ``modified_after`` - Zulu ISO 8601 format e.g. ``modified_after=2016-08-22T09:40:07Z``
- ``modified_before`` - Zulu ISO 8601 format e.g. ``modified_before=2016-08-23T09:40:07Z``
- ``name``
- ``new_repo_task_id``
- ``owner``
- ``rebuild_strategy``
- ``reuse_components_from`` - the compatible module that was used for component reuse
- ``scmurl``
- ``state`` - Can be the state name or the state ID e.g. ``state=done``. This
  parameter can be given multiple times, in which case or-ing will be used.
- ``state_reason``
- ``stream``
- ``stream_version_lte`` - less than or equal to the stream version. This is limited to
  the major version. This value only applies to base modules.
- ``submitted_after`` - Zulu ISO 8601 format e.g. ``submitted_after=2016-08-22T09:40:07Z``
- ``submitted_before`` - Zulu ISO 8601 format e.g. ``submitted_before=2016-08-23T09:40:07Z``
- ``version``

An example of querying the "module-builds" resource with the "state",
and the "submitted_before" parameters::

    GET /module-build-service/1/module-builds/?state=done&submitted_before=2016-08-23T08:10:07Z

::

    HTTP 200 OK

::

    {
      "items": [
        {
          "id": 3,
          "state": 3,
          ...
        },
        {
          "id": 2,
          "state": 3,
          ...
        },
        {
          "id": 1,
          "state": 3,
          ...
        }
      ],
      "meta": {
        "first": "https://127.0.0.1:5000/module-build-service/1/module-builds/?per_page=10&page=1",
        "last": "https://127.0.0.1:5000/module-build-service/1/module-builds/?per_page=10&page=1",
        "page": 1,
        "pages": 1,
        "per_page": 3,
        "total": 3
      }

Component build state query
---------------------------

Getting particular component build is very similar to a module build query.

::

    GET /module-build-service/1/component-builds/1

The response, if the build exists, would include various pieces of information
about the referenced component build.

::

    HTTP 200 OK

::

    {
      "format": "rpms",
      "id": 854,
      "module_build": 42,
      "nvr": "pth-1-1",
      "package": "pth",
      "state": 1,
      "state_name": "COMPLETE",
      "state_reason": "",
      "task_id": 18367215
    }


The response includes:

- ``id`` - the ID of the component build.
- ``format`` - typically "rpms".
- ``nvr`` - the NVR of the component build.
- ``package`` - the package name.
- ``state`` - the numerical state of the component build.
- ``state_name`` - the named component build state and can be "COMPLETE",
  "FAILED", or "CANCELED".
- ``state_reason`` - the reason why the component build is in this state. This is useful
  when the build fails.
- ``task_id`` - the related task ID in the backend buildsystem.


Listing component builds
------------------------

An example of querying the "component-builds" resource without any additional
parameters::

    GET /module-build-service/1/component-builds/

::

    HTTP 200 OK

::

    {
      "items": [
        {
          "format": "rpms",
          "id": 854,
          "module_build": 42,
          "package": "pth",
          "state": 1,
          "state_name": "COMPLETE",
          "state_reason": "",
          "state_trace": [
            {
              "reason": "Submitted pth to Koji",
              "state": 0,
              "state_name": "init",
              "time": "2017-03-14T00:07:43Z"
            },
            {
              "reason": "",
              "state": 1,
              "state_name": "wait",
              "time": "2017-03-14T00:13:30Z"
            },
            {
              "reason": "",
              "state": 1,
              "state_name": "wait",
              "time": "2017-03-14T14:41:21Z"
            }
          ],
          "task_id": 18367215
        },
        ...
      ],
      "meta": {
        "first": "http://mbs.fedoraproject.org/module-build-service/1/component-builds/?per_page=10&page=1",
        "last": "http://mbs.fedoraproject.org/module-build-service/1/component-builds/?per_page=10&page=5604",
        "next": "http://mbs.fedoraproject.org/module-build-service/1/component-builds/?per_page=10&page=2",
        "page": 1,
        "pages": 5604,
        "per_page": 10,
        "prev": null,
        "total": 56033
      }
    }



Filtering component builds
--------------------------

The component builds can be filtered by a variety of GET parameters. Some of these
parameters include:

- ``batch``
- ``build_time_only`` - boolean e.g. "true" or "false"
- ``format``
- ``module_id`` or ``module_build``
- ``nvr``
- ``package``
- ``ref``
- ``scmurl``
- ``state`` - Can be the state name or the state ID. Koji states are used
  for resolving to IDs. This parameter can be given multiple times, in which
  case or-ing will be used.
- ``state_reason``
- ``tagged`` - boolean e.g. "true" or "false"
- ``tagged_in_final`` - boolean e.g. "true" or "false"
- ``task_id``


Import module
-------------

Importing of modules is done via posting the SCM URL of a repository
which contains the generated modulemd YAML file. Name, stream, version,
context and other important information must be present in the metadata.

::

    POST /module-build-service/1/import-module/

::

    {
      "scmurl": "https://src.fedoraproject.org/modules/foo.git?#21f92fb05572d81d78fd9a27d313942d45055840"
    }


If the module build is imported successfully, JSON containing the most
important information is returned from MBS. The JSON also contains log
messages collected during the import.

::

    HTTP 201 Created

::

    {
      "module": {
        "component_builds": [],

docs/CHANGELOG.rst
Change Log
==========

v3.4.1
------
* Do not add conflicts in dependencies when builds are identical

v3.4.0
------
* Implement static contexts in stream expansion

v3.3.0
------
* Get latest build of latest stream when determining requirements

v3.2.1
------
* Fix a bug with final modulemd call

v3.2.0
------
* Add API call to get final modulemd files of builds

v3.1.0
------
* New ``module_stream`` optional parameter for scratch module builds
* Use Koji repo IDs instead of symlinks when accessing build artificats
* Add the ability to limit arches for a build via the ``buildopts.arches`` field in modulemd
* Skip git ref checks for rpm components with srpm overrides
* Make ``record_module_build_arches`` idempotent to avoid multiple calls
* Honor the ``MBS_CONFIG_SECTION`` environment variable when performing a local build

v3.0.0
------
* Rearchitect MBS to support multiple backends using Celery
* Send module build state changes only after they are available via the REST API
* Set ``mock.yum.module_hotfixes = 1`` on created Koji build tags by default
* Use Z stream base module streams per the schedule in product pages as well as the
  existing behavior of past the GA date

v2.32.0
-------
* Fix the provides of base modules when they have a stream version
* Use gssapi if correct python-requests-kerberos is available
* Actually fail if dnf can't read the repo
* Make dnf timeout configurable

v2.31.0
-------
* Use jinja templates to provide 'full-jobs'
* Only allow cancelling module builds in the init, wait, and build states
* Remove the koji.ClientSession backport
* Gracefully handle builds without a koji tag
* Allow koji tags to be created with a configurable permission

v2.30.4
-------
* allow component reuse in some cases when a component is added

v2.30.3
-------
* Fix a local build bug caused by the refactoring of how database sessions are handled

v2.30.2
-------
* Fixed bugs that caused local builds to fail on Fedora 31

v2.30.1
-------
* Fixed a bug that caused local builds to fail depending on the version of DNF being used

v2.30.0
-------
* Remove Kerberos authentication directly performed by MBS (``mod_auth_gssapi`` is now required)
* Find the correct default modules based on the buildrequired base module (e.g. platform)
* Fix the name field in the modulemd text for ``-devel`` content generator builds
* The KojiResolver feature is now ready for testing
* Fixed a bug that caused local builds to fail

v2.29.1
-------
* Fix a fork in the database migration scripts

v2.29.0
-------
* Add buildonly support (#1307)
* Make the DNF minrate setting configurable when loading repos
* Load the DNF repos in parallel to improve speed
* Record the build_context without base modules
* Convert arch to Koji's canon arch for default modules
* Initial KojiResolver code

v2.28.2
-------
* Fix a bug that caused the code that handles the conflicts in base module modular Koji tags to not run
* Modifications to the stream when querying Product Pages are now on the GA date and not after

v2.28.1
-------
* Fix the ``list_of_args and list_of_kwargs must be list or None`` error when building a module

v2.28.0
-------
* Do not ignore building components to reset state when submit an existing module build
* Handle the conflicts between base module modular Koji tags everytime
* Fix DetachedInstanceError on local builds
* Fix an issue where components were reused from a module built with an incompatible platform

v2.27.0
-------
* Add "scratch_build_only_branches" configuration options to define the branches
  from which only scratch builds can be submitted.
* Do not check Greenwave gating status for scratch builds.
* Fix invalid date in module-build-macros when building modules locally when non-US
  locale is used.
* Fix --add-local-build with MBS Resolver.
* Fail the module build in case it buildrequires module build which does not have
  all the MSE builds in failed or ready state.
* Fix the issue in component reuse code when it could reuse components from module
  build built against older version of base module.

v2.26.0
-------
* Allow the poller to clean up module builds without arches.
* Prevent overlapping RPMs from buildrequired base modules from being available when using
  default modules.
* Add an REST endpoint to display log messages.
* Allow reusing components from builds built against compatible base module streams.
* Serialize component build state trace correctly if state is unset.
* Return empty result if short=true is specified on empty list of builds.


v2.25.0
-------
* Add support for "default modules" (Ursa-Prime)
* Allow modules built against an EoL platform to be used as a build dependency
* Allow the poller to clean up module builds without arches
* Default the cache directory to "/tmp/mbs" instead of under the home directory
* Do not start a Koji newRepo task if one is in progress
* Support base modules with virtual streams and no stream version
* Allow specifying a specific module build to reuse components from
* Fix a bug that prevented local module builds from completing after the libmodulemd v2 migration


v2.24.0
-------
* Allow configuring the number of parallel submissions to Koji
* Add state_reason on a Greenwave failure
* Module builds now remember what module they reused for building
* Fix type error preventing inclusion of custom SRPMs when running on F31/Rawhide
* Fix decoding issue with ``kobo.rpmlib.get_header_field()``


v2.23.0
-------
* Accept floats when filtering by stream_version_lte on the API.
* When no architecture is set in Koji tag, fallback to ``conf.arches``.


v2.22.0
-------
* Allow configuring ``dynamic_buildrequires``
* Allow buildrequiring modules built against all platform streams
* Take the list of arches for ``-build`` Koji tag from buildrequired modules
* Make ``sync_koji_build_tags`` poller working only with the builds that are in build state
  for some time
* Use single session object in greenwave handler and call ``commit()`` in the end

rida.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# Copyright (c) 2016  Red Hat, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
# Written by Petr Šabata <contyk@redhat.com>

"""The module build orchestrator for Modularity, API.

This is the implementation of the orchestrator's public RESTful API.
"""

from flask import Flask, request
import flask
import json
import logging
import modulemd
import os
import rida.auth
import rida.config
import rida.database
import rida.logger
import rida.scm
import ssl
import shutil
import tempfile

app = Flask(__name__)
app.config.from_envvar("RIDA_SETTINGS", silent=True)

ridaconfig=os.environ.get('RIDA_CONFIG')
if ridaconfig:
    conf = rida.config.from_file(ridaconfig)
else:
    conf = rida.config.from_file("rida.conf")
rida.logger.init_logging(conf)

log = logging.getLogger(__name__)

db = rida.database.Database(conf)

@app.route("/rida/module-builds/", methods=["POST"])
def submit_build():
    """Handles new module build submissions."""

    username = rida.auth.is_packager(conf.pkgdb_api_url)
    if not username:
        return ("You must use your Fedora certificate when submitting"
               " new build", 403)

    try:
        r = json.loads(request.get_data().decode("utf-8"))
    except:
        return "Invalid JSON submitted", 400
    if "scmurl" not in r:
        return "Missing scmurl", 400
    url = r["scmurl"]
    urlallowed = False
    for prefix in conf.scmurls:
        if url.startswith(prefix):
            urlallowed = True
            break
    if not urlallowed:
        return "The submitted scmurl isn't allowed", 403
    yaml = str()
    try:
        td = tempfile.mkdtemp()
        scm = rida.scm.SCM(url, conf.scmurls)
        cod = scm.checkout(td)
        cofn = os.path.join(cod, (scm.name + ".yaml"))
        with open(cofn, "r") as mmdfile:
            yaml = mmdfile.read()
    except Exception as e:
        if "is not in the list of allowed SCMs" in str(e):
            rc = 403
        elif "Invalid SCM URL" in str(e):
            rc = 400
        else:
            rc = 500
        return str(e), rc
    finally:
        shutil.rmtree(td)
    mmd = modulemd.ModuleMetadata()
    try:
        mmd.loads(yaml)
    except:
        return "Invalid modulemd", 422
    if db.session.query(rida.database.ModuleBuild).filter_by(name=mmd.name,
        version=mmd.version, release=mmd.release).first():
        return "Module already exists", 409
    module = rida.database.ModuleBuild.create(
        db.session,
        conf,
        name=mmd.name,
        version=mmd.version,
        release=mmd.release,
        modulemd=yaml,
        scmurl=url,
    )

    def failure(message, code):
        # TODO, we should make some note of why it failed in the db..
        log.exception(message)
        module.transition(conf, rida.database.BUILD_STATES["failed"])
        db.session.add(module)
        db.session.commit()
        return message, code

    for pkgname, pkg in mmd.components.rpms.packages.items():
        if pkg.get("repository") and not conf.rpms_allow_repository:
            return failure("Custom component repositories aren't allowed", 403)
        if pkg.get("cache") and not conf.rpms_allow_cache:
            return failure("Custom component caches aren't allowed", 403)
        if not pkg.get("repository"):
            pkg["repository"] = conf.rpms_default_repository + pkgname
        if not pkg.get("cache"):
            pkg["cache"] = conf.rpms_default_cache + pkgname
        if not pkg.get("commit"):
            try:
                pkg["commit"] = rida.scm.SCM(pkg["repository"]).get_latest()
            except Exception as e:
                return failure("Failed to get the latest commit: %s" % pkgname, 422)
        full_url = pkg["repository"] + "?#" + pkg["commit"]
        if not rida.scm.SCM(full_url).is_available():
            return failure("Cannot checkout %s" % pkgname, 422)
        build = rida.database.ComponentBuild(
            module_id=module.id,
            package=pkgname,
            format="rpms",
            scmurl=full_url,
        )
        db.session.add(build)
    module.modulemd = mmd.dumps()
    module.transition(conf, rida.database.BUILD_STATES["wait"])
    db.session.add(module)
    db.session.commit()
    logging.info("%s submitted build of %s-%s-%s", username, mmd.name,
            mmd.version, mmd.release)
    return flask.jsonify(module.json()), 201


@app.route("/rida/module-builds/", methods=["GET"])
def query_builds():
    """Lists all tracked module builds."""
    return flask.jsonify([{"id": x.id, "state": x.state}
        for x in db.session.query(rida.database.ModuleBuild).all()]), 200


@app.route("/rida/module-builds/<int:id>", methods=["GET"])
def query_build(id):
    """Lists details for the specified module builds."""
    module = db.session.query(rida.database.ModuleBuild).filter_by(id=id).first()
    if module:
        tasks = dict()
        if module.state != "init":
            for build in db.session.query(rida.database.ComponentBuild).filter_by(module_id=id).all():
                tasks[build.format + "/" + build.package] = \
                    str(build.task_id) + "/" + build.state
        return flask.jsonify({
            "id": module.id,
            "state": module.state,
            "tasks": tasks
            }), 200
    else:
        return "No such module found.", 404

def _establish_ssl_context(conf):
    if conf.ssl_enabled == False:
        return None
    # First, do some validation of the configuration
    attributes = (
        'ssl_certificate_file',
        'ssl_certificate_key_file',
        'ssl_ca_certificate_file',
    )
    for attribute in attributes:
        value = getattr(conf, attribute, None)
        if not value:
            raise ValueError("%r could not be found" % attribute)
        if not os.path.exists(value):
            raise OSError("%s: %s file not found." % (attribute, value))

    # Then, establish the ssl context and return it
    ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
    ssl_ctx.load_cert_chain(conf.ssl_certificate_file,
                            conf.ssl_certificate_key_file)
    ssl_ctx.verify_mode = ssl.CERT_OPTIONAL
    ssl_ctx.load_verify_locations(cafile=conf.ssl_ca_certificate_file)
    return ssl_ctx


if __name__ == "__main__":
    logging.info("Starting Rida")
    ssl_ctx = _establish_ssl_context(conf)
    app.run(
        host=conf.host,
        port=conf.port,
        request_handler=rida.auth.ClientCertRequestHandler,
        ssl_context=ssl_ctx,
    )
